Sources/Sentry/Profiling/SentryBacktrace.mm

```diff
@@ -0,0 +1,169 @@
+#else
+#define ptrauth_strip(__value, __key) __value
+#endif
+
+using namespace sentry::profiling;
+using namespace sentry::profiling::thread;
+
+#define LIKELY(x) __builtin_expect(!!(x), 1)
+#define UNLIKELY(x) __builtin_expect(!!(x), 0)
+
+namespace {
+ALWAYS_INLINE bool isValidFrame(std::uintptr_t frame, const StackBounds &bounds) {
+    return bounds.contains(frame) && StackFrame::isAligned(frame);
+}
+
+ALWAYS_INLINE std::uintptr_t stripPtrAuthentication(std::uintptr_t retAddr) {
+    // https://github.com/apple/darwin-xnu/blob/8f02f2a044b9bb1ad951987ef5bab20ec9486310/osfmk/kern/backtrace.c#L120
+    return reinterpret_cast<std::uintptr_t>(
+      ptrauth_strip(reinterpret_cast<void *>(getPreviousInstructionAddress(retAddr)), ptrauth_key_return_address));
+}
+
+constexpr std::size_t kMaxBacktraceDepth = 128;
+
+} // namespace
+
+namespace sentry {
+namespace profiling {
+NOT_TAIL_CALLED NEVER_INLINE std::size_t backtrace(const ThreadHandle &targetThread,
+                                                   const ThreadHandle &callingThread,
+                                                   std::uintptr_t *addresses,
+                                                   const StackBounds &bounds,
+                                                   bool *reachedEndOfStackPtr,
+                                                   std::size_t maxDepth,
+                                                   std::size_t skip) noexcept {
+    assert(addresses != nullptr);
+    if (UNLIKELY(maxDepth == 0 || !bounds.isValid())) {
+        return 0;
+    }
+    std::size_t depth = 0;
+    MachineContext machineContext;
+    if (fillThreadState(targetThread.nativeHandle(), &machineContext) != KERN_SUCCESS) {
+        SENTRY_LOG_ASYNC_SAFE_ERROR("Failed to fill thread state");
+        return 0;
+    }
+    if (LIKELY(skip == 0)) {
+        addresses[depth++] = getProgramCounter(&machineContext);
+    } else {
+        skip--;
+    }
+    if (LIKELY(depth < maxDepth)) {
+        const auto lr = getLinkRegister(&machineContext);
+        if (isValidFrame(lr, bounds)) {
+            if (LIKELY(skip == 0)) {
+                addresses[depth++] = stripPtrAuthentication(lr);
+            } else {
+                skip--;
+            }
+        }
+    }
+    std::uintptr_t current;
+    if (UNLIKELY(callingThread == targetThread)) {
+        current = reinterpret_cast<std::uintptr_t>(__builtin_frame_address(0));
+    } else {
+        current = getFrameAddress(&machineContext);
+    }
+    // Even if this bounds check passes, the frame pointer address could still be invalid if the
+    // thread was suspended in an inconsistent state. The best we can do is to detect these
+    // situations at symbolication time on the server and filter them out -- there's not an easy
+    // architecture agnostic way to detect this on the client without a more complicated stack
+    // unwinding implementation (e.g. DWARF)
+    if (UNLIKELY(!isValidFrame(current, bounds))) {
```

Comment:
Perhaps something @Swatinem or @flub need to be aware?

Reply:
If you try to symbolicate complete mess symbolicator will already tell you.